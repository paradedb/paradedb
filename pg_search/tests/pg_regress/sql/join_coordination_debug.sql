-- Debug test for JOIN coordination
-- This test helps us understand why JOIN coordination is not being triggered

CREATE EXTENSION IF NOT EXISTS pg_search;

-- Create simple test tables
CREATE TABLE test_docs (
    id SERIAL PRIMARY KEY,
    content TEXT
);

CREATE TABLE test_files (
    id SERIAL PRIMARY KEY,
    doc_id INTEGER REFERENCES test_docs(id),
    title TEXT
);

-- Create BM25 indexes
CALL paradedb.create_bm25_test_table(
    table_name => 'test_docs',
    schema_name => 'public'
);

CALL paradedb.create_bm25_test_table(
    table_name => 'test_files',
    schema_name => 'public'
);

CREATE INDEX test_docs_search_idx ON test_docs USING bm25 (id, content) WITH (key_field='id');
CREATE INDEX test_files_search_idx ON test_files USING bm25 (id, doc_id, title) WITH (key_field='id');

-- Insert test data
INSERT INTO test_docs (content) VALUES
    ('technology document'),
    ('science document');

INSERT INTO test_files (doc_id, title) VALUES
    (1, 'tech report'),
    (2, 'science report');

-- Test: Multi-table query with search predicates and LIMIT
-- This should trigger JOIN coordination if our logic is working
EXPLAIN (ANALYZE, BUFFERS, VERBOSE)
SELECT d.id, f.id
FROM test_docs d
JOIN test_files f ON d.id = f.doc_id
WHERE d.content @@@ 'technology' 
  AND f.title @@@ 'report'
LIMIT 5;

-- Cleanup
DROP TABLE test_files;
DROP TABLE test_docs; 
