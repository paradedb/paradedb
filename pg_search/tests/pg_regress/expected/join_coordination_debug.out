-- Debug test for JOIN coordination
-- This test helps us understand why JOIN coordination is not being triggered
CREATE EXTENSION IF NOT EXISTS pg_search;
-- Enable JOIN coordination GUC and verify it works
SET paradedb.enable_join_coordination = true;
-- Create simple test tables
CREATE TABLE test_docs (
    id SERIAL PRIMARY KEY,
    content TEXT
);
CREATE TABLE test_files (
    id SERIAL PRIMARY KEY,
    doc_id INTEGER REFERENCES test_docs(id),
    title TEXT
);
-- Create BM25 indexes
CALL paradedb.create_bm25_test_table(
    table_name => 'test_docs',
    schema_name => 'public'
);
WARNING:  === JOIN PATHLIST CALLBACK ===
WARNING:    jointype: 0
WARNING:    limit_tuples: -1
WARNING:    outer_has_search: false
WARNING:    inner_has_search: false
WARNING:    SKIPPING: Not both relations have search predicates
WARNING:  === JOIN PATHLIST CALLBACK ===
WARNING:    jointype: 0
WARNING:    limit_tuples: -1
WARNING:    outer_has_search: false
WARNING:    inner_has_search: false
WARNING:    SKIPPING: Not both relations have search predicates
WARNING:  The table public.test_docs already exists, skipping.
CALL paradedb.create_bm25_test_table(
    table_name => 'test_files',
    schema_name => 'public'
);
WARNING:  === JOIN PATHLIST CALLBACK ===
WARNING:    jointype: 0
WARNING:    limit_tuples: -1
WARNING:    outer_has_search: false
WARNING:    inner_has_search: false
WARNING:    SKIPPING: Not both relations have search predicates
WARNING:  === JOIN PATHLIST CALLBACK ===
WARNING:    jointype: 0
WARNING:    limit_tuples: -1
WARNING:    outer_has_search: false
WARNING:    inner_has_search: false
WARNING:    SKIPPING: Not both relations have search predicates
WARNING:  The table public.test_files already exists, skipping.
CREATE INDEX test_docs_search_idx ON test_docs USING bm25 (id, content) WITH (key_field='id');
CREATE INDEX test_files_search_idx ON test_files USING bm25 (id, doc_id, title) WITH (key_field='id');
-- Insert test data
INSERT INTO test_docs (content) VALUES
    ('technology document'),
    ('science document');
INSERT INTO test_files (doc_id, title) VALUES
    (1, 'tech report'),
    (2, 'science report');
WARNING:  === JOIN COORDINATION DEBUG: analyze_multi_table_search_scenario called for RTI 1 ===
WARNING:  Range table has 1 entries
WARNING:  Examining RTI 1: rtekind = 0
WARNING:  RTI 1 has BM25 index: table_oid=test_docs, index_oid=test_docs_search_idx
WARNING:    Checking search predicates for RTI 1
WARNING:    RTI 1 has 1 base restrict info entries
WARNING:      Checking restrict info 0
WARNING:    RTI 1 has NO search operators
WARNING:  RTI 1 has BM25 index but NO search predicates
WARNING:  === FINAL RESULT for RTI 1: is_multi_table=false, count=0, beneficial=false ===
WARNING:  Setting multi-table info for RTI 1: multi_table=false, count=0, beneficial=false
WARNING:  === JOIN COORDINATION DEBUG: analyze_multi_table_search_scenario called for RTI 1 ===
WARNING:  Range table has 1 entries
WARNING:  Examining RTI 1: rtekind = 0
WARNING:  RTI 1 has BM25 index: table_oid=test_docs, index_oid=test_docs_search_idx
WARNING:    Checking search predicates for RTI 1
WARNING:    RTI 1 has 1 base restrict info entries
WARNING:      Checking restrict info 0
WARNING:    RTI 1 has NO search operators
WARNING:  RTI 1 has BM25 index but NO search predicates
WARNING:  === FINAL RESULT for RTI 1: is_multi_table=false, count=0, beneficial=false ===
WARNING:  Setting multi-table info for RTI 1: multi_table=false, count=0, beneficial=false
-- Test: Multi-table query with search predicates and LIMIT
-- This should trigger JOIN coordination if our logic is working
EXPLAIN (FORMAT TEXT, COSTS OFF, TIMING OFF)
SELECT d.id, f.id
FROM test_docs d
JOIN test_files f ON d.id = f.doc_id
WHERE d.content @@@ 'technology' 
  AND f.title @@@ 'report'
LIMIT 5;
WARNING:  === JOIN COORDINATION DEBUG: analyze_multi_table_search_scenario called for RTI 1 ===
WARNING:  Range table has 3 entries
WARNING:  Examining RTI 1: rtekind = 0
WARNING:  RTI 1 has BM25 index: table_oid=test_docs, index_oid=test_docs_search_idx
WARNING:    Checking search predicates for RTI 1
WARNING:    RTI 1 has 1 base restrict info entries
WARNING:      Checking restrict info 0
WARNING:    RTI 1 FOUND search operator in restrict info 0
WARNING:  RTI 1 HAS search predicates! Total count now: 1
WARNING:  Examining RTI 2: rtekind = 0
WARNING:  RTI 2 has BM25 index: table_oid=test_files, index_oid=test_files_search_idx
WARNING:    Checking search predicates for RTI 2
WARNING:    RTI 2 has 1 base restrict info entries
WARNING:      Checking restrict info 0
WARNING:    RTI 2 FOUND search operator in restrict info 0
WARNING:  RTI 2 HAS search predicates! Total count now: 2
WARNING:  Examining RTI 3: rtekind = 2
WARNING:  RTI 3 is not a relation, skipping
WARNING:  Found 2 tables with search predicates - marking as beneficial JOINs
WARNING:  === FINAL RESULT for RTI 1: is_multi_table=true, count=2, beneficial=true ===
WARNING:  Setting multi-table info for RTI 1: multi_table=true, count=2, beneficial=true
WARNING:  === CHOOSING EXECUTION METHOD ===
WARNING:    limit: None
WARNING:    sort_direction: None
WARNING:  === CHECKING JOIN COORDINATION ELIGIBILITY ===
WARNING:    is_multi_table_search: true
WARNING:    multi_table_search_count: 2
WARNING:    has_limit: false
WARNING:    limit_value: None
WARNING:    has_beneficial_joins: true
WARNING:    sort_direction: None
WARNING:    DISABLED: JOIN coordination now handled at JOIN planning level
WARNING:    CHOSEN: FastFieldNumeric
WARNING:  === JOIN COORDINATION DEBUG: analyze_multi_table_search_scenario called for RTI 2 ===
WARNING:  Range table has 3 entries
WARNING:  Examining RTI 1: rtekind = 0
WARNING:  RTI 1 has BM25 index: table_oid=test_docs, index_oid=test_docs_search_idx
WARNING:    Checking search predicates for RTI 1
WARNING:    RTI 1 has 1 base restrict info entries
WARNING:      Checking restrict info 0
WARNING:    RTI 1 FOUND search operator in restrict info 0
WARNING:  RTI 1 HAS search predicates! Total count now: 1
WARNING:  Examining RTI 2: rtekind = 0
WARNING:  RTI 2 has BM25 index: table_oid=test_files, index_oid=test_files_search_idx
WARNING:    Checking search predicates for RTI 2
WARNING:    RTI 2 has 1 base restrict info entries
WARNING:      Checking restrict info 0
WARNING:    RTI 2 FOUND search operator in restrict info 0
WARNING:  RTI 2 HAS search predicates! Total count now: 2
WARNING:  Examining RTI 3: rtekind = 2
WARNING:  RTI 3 is not a relation, skipping
WARNING:  Found 2 tables with search predicates - marking as beneficial JOINs
WARNING:  === FINAL RESULT for RTI 2: is_multi_table=true, count=2, beneficial=true ===
WARNING:  Setting multi-table info for RTI 2: multi_table=true, count=2, beneficial=true
WARNING:  === CHOOSING EXECUTION METHOD ===
WARNING:    limit: None
WARNING:    sort_direction: None
WARNING:  === CHECKING JOIN COORDINATION ELIGIBILITY ===
WARNING:    is_multi_table_search: true
WARNING:    multi_table_search_count: 2
WARNING:    has_limit: false
WARNING:    limit_value: None
WARNING:    has_beneficial_joins: true
WARNING:    sort_direction: None
WARNING:    DISABLED: JOIN coordination now handled at JOIN planning level
WARNING:    CHOSEN: FastFieldNumeric
WARNING:  === JOIN PATHLIST CALLBACK ===
WARNING:    jointype: 0
WARNING:    limit_tuples: 5
WARNING:    outer_has_search: true
WARNING:    inner_has_search: true
WARNING:    has_limit: true
WARNING:    CREATING JOIN COORDINATION PATH!
WARNING:      RTI 1 maps to table OID test_docs
WARNING:      Found table OID test_docs for RTI 1
WARNING:    Extracted outer table OID: test_docs (RTI: 1)
WARNING:      RTI 2 maps to table OID test_files
WARNING:      Found table OID test_files for RTI 2
WARNING:    Extracted inner table OID: test_files (RTI: 2)
WARNING:      RTI 1 maps to table OID test_docs
WARNING:      RTI 2 maps to table OID test_files
WARNING:    Extracted JOIN condition: test_docs[1] = test_files[2]
WARNING:    Extracted 1 JOIN conditions
WARNING:    JOIN private data: 2 tables, limit: Some(5)
WARNING:    Serialized JSON: {"relation_mapping":{"164187":1,"164196":2},"table_oids":[164187,164196],"limit":5,"join_conditions":[{"left":[164187,1],"right":[164196,2],"opno":96}]}
WARNING:    Created persistent string at: 0x7ff6e88c4f08
WARNING:    Created string node: 0x7ff6e88c5018
WARNING:    Added node to list, list length: 1
WARNING:    Creating custom path: startup_cost=20, total_cost=1.6, rows=1
WARNING:    CustomPath created successfully with joinrel.reltarget
WARNING:    SUCCESSFULLY CREATED JOIN COORDINATION PATH!
WARNING:  === JOIN PATHLIST CALLBACK ===
WARNING:    jointype: 0
WARNING:    limit_tuples: 5
WARNING:    outer_has_search: true
WARNING:    inner_has_search: true
WARNING:    has_limit: true
WARNING:    CREATING JOIN COORDINATION PATH!
WARNING:      RTI 2 maps to table OID test_files
WARNING:      Found table OID test_files for RTI 2
WARNING:    Extracted outer table OID: test_files (RTI: 2)
WARNING:      RTI 1 maps to table OID test_docs
WARNING:      Found table OID test_docs for RTI 1
WARNING:    Extracted inner table OID: test_docs (RTI: 1)
WARNING:      RTI 1 maps to table OID test_docs
WARNING:      RTI 2 maps to table OID test_files
WARNING:    Extracted JOIN condition: test_docs[1] = test_files[2]
WARNING:    Extracted 1 JOIN conditions
WARNING:    JOIN private data: 2 tables, limit: Some(5)
WARNING:    Serialized JSON: {"relation_mapping":{"164187":1,"164196":2},"table_oids":[164196,164187],"limit":5,"join_conditions":[{"left":[164187,1],"right":[164196,2],"opno":96}]}
WARNING:    Created persistent string at: 0x7ff6e88c7628
WARNING:    Created string node: 0x7ff6e88c7738
WARNING:    Added node to list, list length: 1
WARNING:    Creating custom path: startup_cost=20, total_cost=1.6, rows=1
WARNING:    CustomPath created successfully with joinrel.reltarget
WARNING:    SUCCESSFULLY CREATED JOIN COORDINATION PATH!
WARNING:  === PLANNING JOIN COORDINATION CUSTOM PATH ===
WARNING:    rel relids: ["test_docs", "test_files"]
WARNING:    tlist length: 2
WARNING:    Private data: 2 tables, limit: Some(5)
WARNING:    Using joinrel target list with 2 expressions
WARNING:    Created CustomScan plan for JOIN coordination
WARNING:    startup_cost: 20, total_cost: 1.6, rows: 1
ERROR:  variable not found in subplan target list
-- Cleanup
DROP TABLE test_files;
DROP TABLE test_docs; 
RESET paradedb.enable_join_coordination;
